File: go-mqtt-backend/internal/logger/logger.go
// Package logger 提供全局 zap 结构化日志实例。
// 对应 bun-mqtt-backend/src/logger.ts 的 Pino 日志。
package logger

import (
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

// Log 全局日志实例，由 Init() 初始化后供其他包使用。
var Log *zap.Logger = zap.NewNop()

// Init 根据日志级别初始化全局 Logger。
// level: debug / info / warn / error / fatal
func Init(level string) {
	var lvl zapcore.Level
	switch level {
	case "debug":
		lvl = zapcore.DebugLevel
	case "info":
		lvl = zapcore.InfoLevel
	case "warn":
		lvl = zapcore.WarnLevel
	case "error":
		lvl = zapcore.ErrorLevel
	case "fatal":
		lvl = zapcore.FatalLevel
	default:
		lvl = zapcore.InfoLevel
	}

	cfg := zap.Config{
		Level:            zap.NewAtomicLevelAt(lvl),
		Encoding:         "json",
		OutputPaths:      []string{"stdout"},
		ErrorOutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			TimeKey:        "time",
			LevelKey:       "level",
			NameKey:        "app",
			MessageKey:     "msg",
			CallerKey:      "",
			StacktraceKey:  "stacktrace",
			LineEnding:     zapcore.DefaultLineEnding,
			EncodeLevel:    zapcore.LowercaseLevelEncoder,
			EncodeTime:     zapcore.ISO8601TimeEncoder,
			EncodeDuration: zapcore.MillisDurationEncoder,
		},
	}
	cfg.InitialFields = map[string]any{"app": "go-mqtt-backend"}

	logger, err := cfg.Build()
	if err != nil {
		panic("logger init failed: " + err.Error())
	}
	Log = logger
}

// Sync 刷新日志缓冲区（程序退出前调用）。
func Sync() {
	_ = Log.Sync()
}

File: rust-mqtt-backend/src/logger.rs
//! 结构化日志初始化（对齐 go-mqtt-backend/internal/logger 的 zap JSON）。
//!
//! ⚠️ 全局 subscriber 只能设一次；main 里最先调用 init(&cfg.log_level)。

use tracing_subscriber::EnvFilter;

/// 按日志级别初始化全局 tracing subscriber（JSON 格式，输出 stdout）。
///
/// level ∈ {debug,info,warn,error,fatal}（fatal 映射到 error，tracing 无 fatal 级）。
pub fn init(level: &str) {
    let filter = match level {
        "debug" => "debug",
        "warn" => "warn",
        "error" | "fatal" => "error",
        _ => "info",
    };

    let _ = tracing_subscriber::fmt()
        .json()
        .with_target(false)
        .flatten_event(true)
        .with_current_span(false)
        .with_span_list(false)
        .with_env_filter(EnvFilter::new(filter))
        .try_init();
}

/// 刷新日志缓冲（tracing fmt 无显式缓冲，此函数为对齐 Go logger.Sync() 而保留）。
pub fn sync() {
    // tracing-subscriber 的 fmt 层直接写 stdout，无需手动 flush。
}
